Crayfish, Lobster Animations, and Cougars

Therighthon opened the pull request:
Added Freshwater Critters as an extension of Aquatic Critters
Added a crayfish as the sole variety of Freshwater critter
Made critters escape from the player when attacked
Added a swimming animation to lobsters and crayfish

Added Cougars/Framework for big cats
- Big cats extend predators, and have a stalking behavior for players similar to how vanilla ocelots stalk chickens.
- Black panthers implemented as a rare variant of cougar.
- Did a lot of experimentation to make walk animations nice and smooth when using animations from block bench.

Therighthon committed:
Added Crayfish

Added FreshWaterCritters, and Crayfish as the sole variant.

Therighthon committed:
Getting Crayfish to Spawn

Therighthon committed:
Finished Lobster/Crayfish Implementation

Separate Swimming Animation
Both now use the LobsterModel correctly
Aquatic critters flee when attacked

Therighthon committed:
Cougars

Added cougars, a new predator type that will attempt to sneak up on the player before attacking.

Therighthon committed:
Added Panthers

Added panthers as a rare (1 in 10) variant of the cougar

Therighthon committed:
Nocturnal Cougars

Set cougars to be nocturnal

eerussianguy commented on src/main/java/net/dries007/tfc/client/model/entity/LobsterModel.java:
```diff
@@ -23,50 +30,68 @@ public static LayerDefinition createBodyLayer()
         return LayerDefinition.create(meshdefinition, 32, 32);
     }
+
+    //Animation transubstantiated by Troggacompany
```
lets not put credits like this inside the code, that's what commit histories are for

eerussianguy commented on src/main/java/net/dries007/tfc/client/model/entity/LobsterModel.java:
```diff
@@ -78,27 +103,43 @@ public LobsterModel(ModelPart root)
+        {
+            final float adjustedAgeInTicks = ageInTicks + limbSwingAmount * 2F;
+            SWIM.tick(parts, adjustedAgeInTicks);
+        } else {
```
code style, brackets

eerussianguy commented on src/main/java/net/dries007/tfc/client/render/entity/CougarRenderer.java:
```diff
@@ -0,0 +1,41 @@
+    @Override
+    protected void scale(Cougar cougar, PoseStack poseStack, float ticks)
+    {
+        poseStack.scale(1.2F, 1.2F, 1.2F);
```
i would prefer the actual mdoel is fixed here rather than hardcoding this but it doesn't matter that much

eerussianguy replied:
further, there should after that be no Cougar class, just the regular feline class that gets instantiated twice

Therighthon replied:
Okay, I think this is just something I copied over from the polar bear class without understanding it, but I can look into changing it.

As far as the Cougar class, you are saying I should just add cougars and panthers as two separate entity types rather than one with two variants? Or are you saying the variant code should be in the BigCat class somehow? I just did it this way because I wasn't sure of a good way to make the panther variant rare if they were separate mobs. And while it isn't that important for panthers to be rare, I am kinda attached to the idea.

eerussianguy replied:
They should be different entity types. We can work on rarity at another point (there's a lot of levers to pull)

eerussianguy commented on src/main/java/net/dries007/tfc/common/entities/predator/BigCat.java:
```diff
@@ -0,0 +1,157 @@
+import javax.annotation.Nullable;
+import java.util.Optional;
+
+public class BigCat extends Predator {
```
FelinePredator

eerussianguy commented on src/main/java/net/dries007/tfc/common/entities/predator/BigCat.java:
```diff
@@ -0,0 +1,157 @@
+        this.attackDistanceSquared = attackDistanceSquared;
+    }
+
+    protected void registerGoals() {
+        this.goalSelector.addGoal(7, new LeapAtTargetGoal(this, 0.4F));
+        this.goalSelector.addGoal(8, new BigCatAttackGoal(this));
+        this.targetSelector.addGoal(1, new BigCatNearestAttackableTargetGoal<>(this, Player.class, false));
+        this.targetSelector.addGoal(1, new BigCatNearestAttackableTargetGoal<>(this, Pig.class, false));
+        super.registerGoals();
```
Yeah so we are not going to combine goal AI and brain AI. the predator implementation specifically uses Brain (see that class and PredatorAI for details) if you do this you're overriding basically all of that. So this needs to go

eerussianguy commented on src/main/java/net/dries007/tfc/common/entities/predator/Cougar.java:
```diff
@@ -0,0 +1,85 @@
+    private static final EntityDataAccessor<Integer> DATA_ID_TYPE_VARIANT = SynchedEntityData.defineId(Cougar.class, EntityDataSerializers.INT);
+
+    private static final ResourceLocation[] LOCATIONS = {
+        RenderHelpers.animalTexture("cougar"),
+        RenderHelpers.animalTexture("panther")
```
This should be Helpers#animalTexture

eerussianguy commented on src/main/java/net/dries007/tfc/common/entities/predator/Cougar.java:
```diff
@@ -0,0 +1,85 @@
+import net.minecraft.world.level.ServerLevelAccessor;
+import org.jetbrains.annotations.Nullable;
+
+public class Cougar extends BigCat {
```
Black panthers should be a separate entity type rather than this special implementation, that's why we have the ability to do the `createDiurnal` thing and all that to make instantiation easy

eerussianguy commented on venv/Lib/site-packages/numpy/core/include/numpy/libdivide/LICENSE.txt:
```diff
@@ -0,0 +1,21 @@
+  zlib License
+  ------------
+
+  Copyright (C) 2010 - 2019 ridiculous_fish, <libdivide@ridiculousfish.com>
+  Copyright (C) 2016 - 2019 Kim Walisch, <kim.walisch@gmail.com>
+
+  This software is provided 'as-is', without any express or implied
+  warranty.  In no event will the authors be held liable for any damages
```
remove this from git

eerussianguy commented on venv/Lib/site-packages/pip-22.0.4.dist-info/LICENSE.txt:
```diff
@@ -0,0 +1,20 @@
+distribute, sublicense, and/or sell copies of the Software, and to
+permit persons to whom the Software is furnished to do so, subject to
+the following conditions:
+
```
remove this from git

eerussianguy commented on venv/Lib/site-packages/mcresources-1.5.7.dist-info/LICENSE.txt:
```diff
@@ -0,0 +1,21 @@
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
```
remove this from git

eerussianguy commented on venv/Lib/site-packages/numpy-1.22.3.dist-info/LICENSE.txt:
```diff
@@ -0,0 +1,968 @@
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions are
+met:
+
+    * Redistributions of source code must retain the above copyright
+       notice, this list of conditions and the following disclaimer.
+
+    * Redistributions in binary form must reproduce the above
+       copyright notice, this list of conditions and the following
+       disclaimer in the documentation and/or other materials provided
```
remove this from git

eerussianguy commented on venv/Lib/site-packages/numpy-1.22.3.dist-info/LICENSE.txt:
```diff
@@ -0,0 +1,968 @@
+rights of fair use or other equivalent, as provided by copyright law.
+
+  You may make, run and propagate covered works that you do not
+convey, without conditions so long as your license otherwise remains
+in force.  You may convey covered works to others for the sole purpose
+of having them make modifications exclusively for you, or provide you
+with facilities for running those works, provided that you comply with
+the terms of this License in conveying all material for which you do
```
remove this from git

Therighthon committed:
Addressing minor comments

Therighthon committed:
Delete LICENSE.txt

Therighthon committed:
Delete LICENSE.txt

Therighthon committed:
Delete LICENSE.txt

Therighthon committed:
Delete LICENSE.txt

Therighthon committed:
Delete LICENSE.txt

Therighthon committed:
Removed Cougar Class

eerussianguy commented on src/main/java/net/dries007/tfc/common/entities/predator/FelinePredator.java:
```diff
@@ -1,35 +1,23 @@
-import java.util.Optional;
 
-public class BigCat extends Predator {
+public class FelinePredator extends Predator {
```
need to apply the code style everywhere not just where i commented

Therighthon committed:
Big Cats and Animations

Big Cats:
- Lions, Cougars, Panthers, Sabertooths (Saberteeth?)
- Currently behave like normal predators, lower HP than bears but higher attack

Model Changes
- Polar Bears have walk/run/swim/slide animations for movement.
- Big cats have walk/run animations, cougars/panthers still have a sneak animation, but that is currently unused.
- All predators have sleeping animations rather than just poses
- Made alpacas move their heads and necks independently when looking around.
- Predators rear up when attacking

Added taro root, functionally identical to cattail root, but drops from water taro, an emergent found farther south.

Therighthon committed:
Finally learned how to apply the CodeStyle...

...and applied it to all the Java files in the diff, so should be good to go on that front.

eerussianguy commented on src/main/java/net/dries007/tfc/common/entities/predator/FelinePredator.java:
```diff
@@ -17,128 +17,32 @@
-    public final double attackDistanceSquared;
+public class FelinePredator extends Predator
+{
+    public final double CROUCH_SPEED_MOD;
+    public final double SPRINT_SPEED_MOD;
+    public final double ATTACK_DISTANCE_SQUARED;
```
code style -- these should be casedLikeThis

eerussianguy commented on src/main/java/net/dries007/tfc/common/entities/predator/Predator.java:
```diff
@@ -45,24 +46,37 @@ public static AttributeSupplier.Builder createAttributes()
 
     public final boolean diurnal;
     private int attackAnimationRemainingTicks = 0;
+    public final boolean isMale= random.nextBoolean();
```
bad equals sign

eerussianguy replied:
also you have to serialize this value or you will change the gender of the animal each time its loaded

Therighthon replied:
Okay, so I still don't really understand what's wrong with having an equals sign, but I assume it's a style thing.

Also wasn't sure what it meant to serialize a value, and while I still don't actually know what it means, I now have a working knowledge that allowed me to fix the problem.

Also, while testing to see if I'd fixed the problem, I discovered that predators despawned when you left the chunks since they extend Mob, which I fixed by just setting persistence required. Idk if there's a better way to do that, but it's at least a band-aid

Therighthon committed:
Bears and Bugfixes

Addressed comments.

Added Black and Brown bears.

Made predators not despawn when you walk too far away.

eerussianguy committed:
Merge branch '1.18.x' into April2022

# Conflicts:
#	resources/constants.py
#	resources/data.py
#	src/main/java/net/dries007/tfc/client/render/entity/BearRenderer.java

eerussianguy committed:
small predator fixes

eerussianguy committed:
Delete .gitignore

eerussianguy committed:
last minute cleanups
